[PR-1268] Follow up article on TTH in GHC 9 (#7)
serokell/blog-posts: Typed Template Haskell/article.md

## Typed splices

Just like we had untyped splices such as `$foo`, now we also have typed splices, written as `$$foo`. In this section, we will see how to define and splice typed Template Haskell code.

> We reiterate one point mentioned in our previous article: in GHC 8, the usage of splices may require parentheses or not. For instance, if `foo` came from a qualified import, then you'd need to write `$$(Foo.foo)` instead. In GHC 9, the parser is more relaxed and will not require parentheses in such situations.

Reviewer: ```suggestion
We reiterate one point mentioned in our previous article: before GHC 9, the usage of splices may require parentheses or not. For instance, if `foo` came from a qualified import, then you'd need to write `$$(Foo.foo)` instead. In GHC 9, the parser is more relaxed and will not require parentheses in such situations.
```

An [open ticket](https://gitlab.haskell.org/ghc/ghc/-/issues/10271) exists describing the issue, but if you run into some strange errors, it's a good idea to keep it in mind.

## Compatibility with GHC 9

In GHC 9, the types used by typed Template Haskell were changed according to the [Make Q (TExp a) into a newtype](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0195-code-texp.rst) proposal.

> This proposal changed the typed expression quasi-quoter so it now returns `Quote m => Code m a` instead of the old `Quote m => m (TExp a)` from [`template-haskell-2.17.0.0`](https://hackage.haskell.org/package/template-haskell-2.17.0.0/docs/Language-Haskell-TH.html#t:Code).

Reviewer: I don’t understand the link to th-2.17. I think, `Code` was implemented exactly at the same time as the overloaded bracket.

In other words, what I am saying is that the proposal you are referring indeed changed the type from `Quote m => m (TExp a)` to `Quote m => Code m a`, but also, a the exact same time another proposal changed the type from `Q (TExp a)` to `Quote m => m (TExp a)`, so the reference to 2.17 does not seem relevant, as both changes happened in there.

To sum up, IMHO, actionable points here are:

1. Maybe mention that it used to be `Q (TExp a)` prior to GHC 9.
2. Remove the link to template-haskell? I mean, I genuinely don’t understand what it means, maybe I am missing something.

Author: I've removed the link and used `Q (TExp a)` there.

Reviewer: Ok, but now the connection between the old time and the new type is harder to grasp... at least without any context. Maybe it would be worth it to somehow point out that `m` is what used to be `Q` before the changes?

Author: I elaborated a bit more. I think it should be clearer now.

> If the reader is curious, please take a moment to read our [Typed Template Haskell in GHC 9](https://serokell.io/blog/typed-template-haskell-in-ghc-9) article detailing a migration guide of typed TH code to GHC 9, where we discuss a backward-compatible way to integrate typed Template Haskell code in both GHC 8 and GHC 9.

Reviewer: ```suggestion
If the reader is curious, please take a moment to read our [Typed Template Haskell in GHC 9](./typed-template-haskell-in-ghc-9) article detailing a migration guide of typed TH code to GHC 9, where we discuss a backward-compatible way of integrating typed Template Haskell code in both GHC 8 and GHC 9.
```

I am not sure about absolute vs. relative URLs, what do we normally use? @NaeosPsy? I mean, I am personally a fan of relative URLs, I think we should use them whenever possible, but maybe there are reasons to prefer absolute?

Reviewer 2: We have historically used absolute (read: I have never imagined to use a relative link like that, nor has anyone else before, I think), and I think it would be simpler to leave it at status quo.

Reviewer: Ok, but if we rename the company and change the domain, you’ll have to go fix it all 🙄

> If the reader is curious, please take a moment to read our [Typed Template Haskell in GHC 9](https://serokell.io/blog/typed-template-haskell-in-ghc-9) article detailing a migration guide of typed TH code to GHC 9, where we discuss a backward-compatible way to integrate typed Template Haskell code in both GHC 8 and GHC 9.

Reviewer: > If the reader is curious, please take a moment to read

This is a weird transition from third person to second person imperative.

Author: Changed "the reader is" to "you're".

Thankfully, changing the examples to work with GHC 9 is not difficult. In short, you may replace the parts like `Q (TExp a)` with `Quote m => Code m a` (or `Code Q a`) and the code should compile.

```hs
primesUpTo' :: Quote m => Integer -> Code m [Integer]
primesUpTo' n = [|| primesUpTo n ||]
```

Furthermore, as discussed, the [caveat](#caveat) in the previous section doesn't apply anymore to GHC 9, and you will be able to write the example above.

```hs
>>> mempty' :: Monoid a => Code Q a
... mempty' = [|| mempty ||]

>>> x :: String
... x = id $$mempty'

>>> x
""
```

> Should you need to explicitly convert between the old and new types, you may pattern match directly on `Code` or use the following functions:

Reviewer: ```suggestion
Should you need to explicitly convert between the old and new types, you may pattern match directly on `Code` or use the following functions from [Language.Haskell.TH](https://hackage.haskell.org/package/template-haskell-2.18.0.0/docs/Language-Haskell-TH.html#t:Code):
```
